keras/src/distribution/tensor_parallel/autoconfig.py

```diff
@@ -0,0 +1,262 @@
+from keras.src import layers
+from keras.src.distribution.tensor_parallel.tensor_layout import LayoutMap
+from keras.src.distribution.tensor_parallel.tensor_layout import (
+    split_tensor_for_parallelism,
+)
+
+_split_fn_internal = split_tensor_for_parallelism
+
+
+def _split_rule(device_count, dim):
+    """
+    Creates a sharding rule for a specific dimension.
+
+    Returns a lambda function compatible with LayoutMap that defines
+    how a tensor should be split across the available devices.
+
+    Args:
+        device_count: The total number of devices available for parallelism.
+        dim: The dimension of the tensor to split.
+
+    Returns:
+        callable: A lambda function accepting (tensor, index) that returns the
+        sharded layout.
+    """
+    return lambda x, index: _split_fn_internal(x, index, device_count, dim=dim)
+
+
+def analyze_dense_layer(layer):
+    """
+    Classifies a Dense layer based on its input/output dimensions.
+
+    This function determines if a Dense layer represents an 'up_projection'
+    (expansion) or a 'down_projection' (contraction) based on a heuristic
+    threshold. This classification dictates how the weights are sharded.
+
+    Heuristic:
+        - Expansion: Output dimension > (Input dimension * 1.5)
+        - Contraction: Input dimension > (Output dimension * 1.5)
+
+    Args:
+        layer (keras.layers.Layer): The layer instance to analyze.
+
+    Returns:
+        str: One of 'up_projection', 'down_projection', or 'dense'.
+    """
+    if not isinstance(layer, layers.Dense):
+        return "dense"
+
+    input_dim = None
+    output_dim = None
+
+    if hasattr(layer, "kernel") and layer.kernel is not None:
```

Comment:
The "real" kernel is `layer._kernel` for Dense and EinsumDense layers in Keras.

`kernel` is a property that may modify the kernel before returning it (for LoRa and quantization).

It may not matter for most of the your code, but it will matter if you want the variable and not just the value of the kernel (for instance to apply sharding to the variable itself).